### PR TITLE
recette convention demat

### PR DESCRIPTION
**Pull request by mfo**
WIP... all questions will be asked during a review btween techs

DONE: 
* poc view_components: (maybe a drapper decorator would be nicer to wrap "view logic/view model")? wdyt
* extract presets of internship application in dedicated model, this implementation feels better imo
* rewire UI based on UX feedback (print button, remove toolnote not focus on current user, remove toolnote on legals, enhance buttons based on recent ux feedback, add progress bar on listing of agreements)
* screenshot across "w3c"/a11y/ui tests.

TODO: 
* change custom dates formats

### CODE REVIEW

File: app/libs/builders/internship_agreement_builder.rb [hunk 1]
```diff
@@ -66,7 +66,7 @@ def preprocess_internship_application_paramns(internship_application)
 
     def preprocess_internship_offer_params(internship_offer)
       {
-        organisation_representative_full_name: internship_offer.employer_name,
+        organisation_representative_full_name: Presenters::User.new(internship_offer.employer).full_name,
```
Comment (mfo) [inline app/libs/builders/internship_agreement_builder.rb:69]: @fitchMitch / @pierrax there was an issue here. it bother me and i'm not confident we can keep this stable without proper testing

File: app/models/ability.rb [hunk 2]
```diff
@@ -95,7 +95,7 @@ def school_manager_abilities(user:)
       edit_main_teacher_full_name
       edit_terms_rich_text
```
Comment (mfo) [inline app/models/ability.rb:96]: @fitchMitch / @pierrax ; i refactored this field, forget an occurence, still a green suite. the sign of missing system test

**Review by mfo — COMMENTED** [review]
s self review in order to openn some discussion and share knowledges around this nearly ready epic

File: app/components/internship_agreements/base_component.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
+module InternshipAgreements
```
[the remark this answers is not in the document]
Comment (fitchMitch) [reply app/components/internship_agreements/base_component.rb:1]: https://thepugautomatic.com/2014/03/draper/ discussion quite interesting --> go on using Presenters

File: app/components/internship_agreements/progress_component.rb [hunk 4]
```diff
@@ -0,0 +1,17 @@
+module InternshipAgreements
+  class ProgressComponent < BaseComponent
+    def curret_role(role_attribute_name:)
+      matches = role_attribute_name.to_s.match(/(?<role>.*)_accept_terms/)
+      raise ArgumentError, "#{user.type} does not support accept terms yet " if matches[:role].empty?
+      matches[:role]
```
Comment (fitchMitch) [inline app/components/internship_agreements/progress_component.rb:6]: add an empty line after guard clause ?

File: app/views/dashboard/internship_agreements/_form_footer.html.slim [hunk 5]
```diff
@@ -1,30 +1,23 @@
...
-            | En signant la convention de stage, vous validez les informations que vous avez renseignées au préalable ainsi que celles renseignées par l’élève et l’entreprise en charge du stage de celui-ci.
+            | En enregistrant la convention de stage, vous validez les informations que vous avez renseignées au préalable ainsi que celles renseignées par l’élève et l’entreprise en charge du stage de celui-ci.
 
-            br
-            br
-
-            | Vous reconnaissez avoir lu et accepté les&nbsp;
-
-            = link_to "Conditions Générales d’Utilisation de monstagedetroisieme.fr",
+            / br
```
Comment (fitchMitch) [inline app/views/dashboard/internship_agreements/_form_footer.html.slim:11]: Un commentaire sur la durée de vie de ces ... commentaires ?

File: app/controllers/dashboard/internship_agreements_controller.rb [hunk 6]
```diff
@@ -38,8 +38,8 @@ def update
       internship_agreement_builder.update(instance: InternshipAgreement.find(params[:id]),
                                           params: internship_agreement_params) do |on|
         on.success do |updated_internship_agreement|
-          redirect_to dashboard_internship_agreement_path(updated_internship_agreement),
-                      flash: { success: 'La convention a été signée.' }
+          redirect_to current_user.custom_agreements_path,
+                      flash: { success: 'La convention a été enregistrée.' }
```
Comment (mfo) [inline app/controllers/dashboard/internship_agreements_controller.rb:42]: ajouté le status [attente de qui doit encore signé ou message pr direque c'est pret a imprimer]

Comment (mfo) [reply app/controllers/dashboard/internship_agreements_controller.rb:42]: @fitchMitch ; je croyais avoir mergé un commit associé a ce sujet, j'imagine qu'un des deux controller a ete traité

File: app/components/internship_agreements/progress_component.rb [hunk 7]
```diff
@@ -0,0 +1,18 @@
+module InternshipAgreements
+  class ProgressComponent < BaseComponent
+   def curret_role(role_attribute_name:)
+      matches = role_attribute_name.to_s.match(/(?<role>.*)_accept_terms/)
+      raise ArgumentError, "#{user.type} does not support accept terms yet " if matches[:role].empty?
+      matches[:role]
```
Comment (mfo) [inline app/components/internship_agreements/progress_component.rb:6]:
```suggestion

      matches[:role]
```

File: app/front_assets/bootapp.js [hunk 8]
```diff
@@ -13,8 +13,15 @@ Rails.start();
...
-application.load(definitionsFromContext(context))
+// const context = require.context('controllers', true, /.js$/)
+// application.load(definitionsFromContext(context))
+const context = require.context("controllers", true, /.js$/)
+const context_components = require.context("../components", true, /_controller.js$/)
+application.load(
+  definitionsFromContext(context).concat(
+    definitionsFromContext(context_components)
+  )
+)
```
Comment (mfo) [inline app/front_assets/bootapp.js:24]:
```suggestion
const context = require.context('controllers', true, /.js$/)
application.load(definitionsFromContext(context))
```

File: app/models/user.rb [hunk 9]
```diff
@@ -12,7 +12,7 @@ class User < ApplicationRecord
   before_validation :clean_phone
   after_create :send_sms_token
 
-  # school_managements includes different roles
+  # school_managernagements includes different roles
```
Comment (mfo) [inline app/models/user.rb:15]:
```suggestion
  # school_managements includes different roles
```

File: app/views/dashboard/internship_agreements/_form.slim [hunk 10]
```diff
@@ -9,11 +9,14 @@ tbody
         .col-8= render "layouts/form_errors", resource: resource
 
       fieldset
-        legend.h4.col-7.px-0 Convention relative à l’organisation de séquence d’observation en milieu professionnel
```
Comment (mfo) [inline app/views/dashboard/internship_agreements/_form.slim]: @pierrax le fait d'avoir viré le legend pete les tests d'a11y

**Comment by mfo** [comment]
@pierrax ; fait parti des actions de notre retro. dois je checker ce qui casse ou t'es dessus ?

**Comment by fitchMitch** [comment]
config/sidekiq.yml définit -batches deux fois ...

**Pull request closed without merging**